src/admin/import_cksum.rs

```diff
@@ -0,0 +1,110 @@
+use std::{
+    fs::File,
+    io::{BufRead, BufReader},
+    thread,
+    time::Duration,
+};
+
+use cargo_registry_index::{Repository, RepositoryConfig};
+use diesel::prelude::*;
+use indicatif::{ProgressBar, ProgressIterator, ProgressStyle};
+
+use crate::{
+    admin::dialoguer,
+    db,
+    schema::{crates, dependencies, versions},
+};
+
+#[derive(clap::Parser, Debug, Copy, Clone)]
+#[clap(
+    name = "import-cksum",
+    about = "Import missing fields from git into the database"
+)]
+pub struct Opts {
+    /// Time in milliseconds to sleep between crate updates to reduce database load.
+    #[clap(long)]
+    pause_millis: u64,
+}
+
+pub fn run(opts: Opts) -> anyhow::Result<()> {
+    let conn = db::oneoff_connection().unwrap();
+    println!("fetching git repo");
+    let config = RepositoryConfig::from_environment();
+    let repo = Repository::open(&config)?;
+    repo.reset_head()?;
+    println!("HEAD is at {}", repo.head_oid()?);
+
+    let files = repo.get_files_modified_since(None)?;
+    println!("found {} crates", files.len());
+    if !dialoguer::confirm("continue?") {
+        return Ok(());
+    }
+
+    let pb = ProgressBar::new(files.len() as u64);
+    pb.set_style(ProgressStyle::with_template("{bar:60} ({pos}/{len}, ETA {eta})").unwrap());
+
+    for file in files.iter().progress_with(pb) {
+        thread::sleep(Duration::from_millis(opts.pause_millis));
+
+        let crate_name = file.file_name().unwrap().to_str().unwrap();
+        let path = repo.index_file(crate_name);
+        if !path.exists() {
+            continue;
+        }
+        let file = File::open(path)?;
+        let reader = BufReader::new(file);
+        for line in reader.lines() {
+            let krate: cargo_registry_index::Crate = serde_json::from_str(&line?)?;
+            conn.transaction(|| import_data(&conn, krate))?;
+        }
+    }
+
+    Ok(())
+}
+
+fn import_data(conn: &PgConnection, krate: cargo_registry_index::Crate) -> QueryResult<()> {
+    let (version_id, checksum): (i32, Option<String>) = versions::table
+        .inner_join(crates::table)
+        .filter(crates::name.eq(&krate.name))
+        .filter(versions::num.eq(&krate.vers))
+        .select((versions::id, versions::checksum))
+        .first(conn)?;
+
+    if checksum.is_none() {
```

Comment:
If the change to add `checksum` is deployed before the change to add `links`, then we'll need to do something different here. Otherwise it may miss importing `links` for crates that already have checksums.

This check assumes that `checksum.is_none()` means that we haven't done the import on the crate yet.

Reply:
yeah, I noticed that. the checksum field has been in prod for a few days already. I guess we can probably clone the index and just go through the files one-by-one to do a full sync with the database. not sure how long this will take but I assume it would faster than the full S3 index sync.